### PR TITLE
[ML] Handle field mismatch in incremental training

### PR DESCRIPTION
**Pull request by tveasey**
The data which was prepared for incremental training of the DGA model from sunburst domains used a different field order and renamed one of the fields. Whilst this wouldn't typically happen in Java, which sorts the field values it supplies by name, it is fragile to rely on this. It is also conceivable that a subset of features aren't available for the update data. This checks that there are no additional fields supplied (likely to indicate a user error) and handles the case that the fields are supplied in a different order or are missing altogether.

### CODE REVIEW

File: lib/api/CDataFrameAnalyzer.cc [hunk 1]
```diff
@@ -250,10 +255,49 @@ void CDataFrameAnalyzer::captureFieldNames(const TStrVec& fieldNames) {
...
+            if (extraColumnNames.empty() == false) {
+                HANDLE_FATAL(<< "Input error: supplying additional columns "
+                             << core::CContainerPrinter::print(extraColumnNames) << ".");
+            }
+
+            for (const auto& name : m_DataFrame->columnNames()) {
+                auto i = std::lower_bound(columnNames.begin(), columnNames.end(), name);
+                if (i == columnNames.end() || *i != name) {
+                    LOG_WARN(<< "Missing column '" << name << "'");
+                    m_ColumnMap->push_back(columnNames.size());
```
Comment (dimitris-athanasiou) [inline lib/api/CDataFrameAnalyzer.cc:291]: This seems to denote missing values with `columnNames.size()`. However, in a comment in `CDataFrameAnalyzer.h` we have ```// This has values: -2 (unset), -1 (missing), >= 0 (control field index).```. Should we not be denoting missing values with `-1`? Or this refers to something else?

File: lib/maths/CBoostedTreeImpl.cc [hunk 2]
```diff
@@ -231,7 +231,7 @@ void CBoostedTreeImpl::train(core::CDataFrame& frame,
 
     m_TrainingProgress.progressCallback(m_Instrumentation->progressCallback());
 
-    std::int64_t lastMemoryUsage(this->memoryUsage());
+    std::int64_t lastMemoryUsage{static_cast<std::int64_t>(this->memoryUsage())};
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTreeImpl.cc:234]: The changes in this file seem unrelated. Are they intentional? (just checking they didn't pop here from another piece of work).

File: lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc [hunk 3]
```diff
@@ -1523,6 +1519,255 @@ BOOST_AUTO_TEST_CASE(testClassificationIncrementalTraining) {
             row->copyTo(newTrainingData.back().begin());
         }
     });
+
+    frame->resizeRows(0);
```
Comment (dimitris-athanasiou) [inline lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:1523]: This seems to complete this test but it's not strictly related to the field order stuff, right?

Comment (tveasey) [reply lib/api/unittest/CDataFrameAnalyzerTrainingTest.cc:1523]: This is just the diff getting confused because there is similarity to test I added. If you scroll down you'll see that this was essentially as it was before.

File: lib/maths/CBoostedTreeImpl.cc (hunk 2 above)
Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:234]: I noticed some compile warnings from these statements when I added some debug. While I could factor this out into a separate change, since it is all on a branch which we'll squash when we merge I'm not sure it is worthwhile.

File: lib/api/CDataFrameAnalyzer.cc (hunk 1 above)
Comment (tveasey) [reply lib/api/CDataFrameAnalyzer.cc]: I agree it would be more consistent in `CDataFrameAnalyzer`, but I also think it is more natural where it's used in `CDataFrame::parseAndWriteRow` to make this a vector of unsigned ints and denote missing as I have done. I guess this is a tradeoff and I don't have strong opinions, if you find the suggested version clearer I'm happy to change: see [e7be3e6](https://github.com/elastic/ml-cpp/pull/2052/commits/e7be3e6d2ddfdd751392090d3f6bed2585b11fce).

Comment (dimitris-athanasiou) [reply lib/api/CDataFrameAnalyzer.cc]: I find it easier to read and understand without e7be3e6. Let it be inconsistent :-)

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM
